### PR TITLE
Feature/reference grid

### PR DESCRIPTION
**Pull request by apmccartney**

### CODE REVIEW

File: src/RECONR/ReferenceGrid/src/call.hpp [hunk 1]
```diff
@@ -1,65 +1,172 @@
...
-  // energies.push_back( range.EL() );
-  // energies.push_back( range.EH() );
+  const auto lowerLimit = caseC.EL();
+  const auto upperLimit = nudgeDown( caseC.EH() );
+
+  std::vector< double > firstPass;
+  firstPass.push_back( lowerLimit );
+
+  {
+    auto jValues = caseC.lValues() | ranges::view::for_each( *this );
```
Comment (apmccartney) [inline src/RECONR/ReferenceGrid/src/call.hpp]: Concatenating the filtered jvalue ranges together then sorting is naive from a complexity point of view, given each jvalue range is already sorted. Merging and sorting using a prioriity queue would be preferable.

That said, I wouldn't expect much of a difference, practically speaking, after accounting for pdqsort's special considerations for mostly sorted ranges.

Comment (apmccartney) [reply src/RECONR/ReferenceGrid/src/call.hpp]: Note that NJOY2016 uses the concatenate and sort approach as well, although there are some important differences. 

1. NJOY2016 eagerly fills each jvalues, which, given the predetermined fill points, amounts to a lot of redunant points.
1. While this sorts at each resonance range (enabling a cheap sort-free merge at the end), NJOY2016 defers the sorting step until after the file3, resolved, and unresolved ranges are concatenated.
1. Here we're using pdqsort, which has `O(N)` complexity for mostly sorted lists. NJOY2016 is using selection sort, which has `O(N^2)` complexity in the best case. 
1. We're using a seperately allocated block of memory for each range whereas NJOY2016 uses a single (fixed size) allocation. This costs a bit more up front but more than pays for the extra allocation time come merge time.

Points [2] and [3] (and to a lesser extent [1] ) taken together are a bit troubling, given, in NJOY2016, sorting has been defered as long as possible. `O(N^2)` gets expensive fast. Might be worthwhile to take a RECONR call on U-238 into gprof and see how much time is spent in the `order` subroutine.

File: src/RECONR/ReferenceGrid/src/fill.hpp [hunk 2]
```diff
@@ -0,0 +1,14 @@
+static void fill( const double lowerBound,
+                  const double upperBound,
+                  std::vector<double>& buffer ){
+  constexpr double thresholdRatio = 1.193776641714437;
```
Comment (jlconlin) [inline src/RECONR/ReferenceGrid/src/fill.hpp]: `<morgan><!-- i.e., physiscist -->`That last digit, `7`, should be a `4`.`</morgan>`
`<kent><!-- i.e., engineer -->`We don't care about that much accuracy.`</kent>`
`<jeremy>`Put a comment at the end saying where that number came from.`</jeremy>`

File: src/RECONR/ReferenceGrid/src/call.hpp [hunk 3]
```diff
@@ -1,65 +1,172 @@
...
+                                   + resonance.GFA()
+                                   + resonance.GFB() );
+  return {{ energy - halfWidth, energy, energy + halfWidth }};
+}
 
-  const double E = resonance.ER();
-  const double hw = resonance.GT()*0.5;
-  return {{E-hw, E, E+hw}};
+template< typename ResolvedLvalue,
+          if_< hasResonances< ResolvedLvalue > > = true >
```
Comment (jlconlin) [inline src/RECONR/ReferenceGrid/src/call.hpp]: Is this `if_` used to help the compiler know what call operator to use?

File: src/RECONR/ReferenceGrid/src/call.hpp [hunk 4]
```diff
@@ -1,65 +1,172 @@
...
-                         resonance.GFB() );
-  buffer = {{E-hw, E, E+hw}};
-  return ranges::make_iterator_range( buffer.begin(), buffer.end() );
+  energies.push_back( upperLimit );
+
+  if ( not ranges::is_sorted( energies ) ){
+    orlp::pdqsort( energies.begin(), energies.end() );
+  }
+
+  energies |= ranges::action::unique;
```
Comment (jlconlin) [inline src/RECONR/ReferenceGrid/src/call.hpp]: I'm not sure what this syntax does. Does it just ensure that all the values are unique? What does the `|=` syntax doe?

File: src/RECONR/ReferenceGrid/src/call.hpp [hunk 5]
```diff
@@ -1,65 +1,172 @@
...
 }
 
-template< typename resolvedL >
-auto operator()( const resolvedL& lValue,
-                 std::conditional_t< true, void*,  decltype( lValue.resonances() ) > = nullptr) const {
+auto operator()( const unresolved::CaseA& caseA ) const {
+  std::vector< double > energies;
+  const auto lowerLimit = caseA.EL();
+  const auto upperLimit = nudgeDown( caseA.EH() );
+  energies.reserve( std::ceil( 5.645828264742274 *
```
Comment (jlconlin) [inline src/RECONR/ReferenceGrid/src/call.hpp]: What is the 5.64... number? Without context, it looks arbitrary.

File: src/RECONR/ReferenceGrid/src/fill.hpp [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  if ( lowerBound * thresholdRatio < upperBound ){
```
Comment (jlconlin) [inline src/RECONR/ReferenceGrid/src/fill.hpp]: I think I know what this is trying to do (add logarithmically spaced energy values between `lowerBound` and `upperBound`), but I can't quite follow what is happening. The syntax is a bit obscure.

File: src/RECONR/ReferenceGrid/test/call.test.cpp [hunk 7]
```diff
@@ -272,3 +241,16 @@ std::string ReichMoore(){
     " 6.209400+0 3.000000+0 1.654700-4 4.005100-2-1.079400-1 7.385900-26922 2151     \n";
 }
 
+/*
```
Comment (jlconlin) [inline src/RECONR/ReferenceGrid/test/call.test.cpp]: Is this test no longer valid or was it accidentally commented out?

File: src/RECONR/ReferenceGrid/src/call.hpp (hunk 3 above)
Comment (apmccartney) [reply src/RECONR/ReferenceGrid/src/call.hpp]: You can see the definition [here](src/RECONR/ReferenceGrid/meta.hpp)

File: src/RECONR/ReferenceGrid/src/call.hpp (hunk 4 above)
Comment (apmccartney) [reply src/RECONR/ReferenceGrid/src/call.hpp]: you can think of this as short hand for
```cpp
energies = std::move(energies) | range::action::unique;
``
which is short for
```cpp
energies = range::action::unique( std::move(energies) );
``
which is short for

```cpp
energies.erase( std::unique( energies.begin(), energies.end() ), energies.end() );
```

which takes an sorted list and removes duplicate entries.

File: src/RECONR/ReferenceGrid/src/call.hpp (hunk 5 above)
Comment (apmccartney) [reply src/RECONR/ReferenceGrid/src/call.hpp]:
```
log_1.193776641714437( EH / EL ) 
= log( EH / EL ) / log( 1.193776641714437 ) 
= log( EH / EL ) / log( 10^(1/13) )
= 13 * log( EH / EL ) / log( 10 )
= 5.645828264742274 * log( EH / EL )
```

I'll add a comment

File: src/RECONR/ReferenceGrid/src/fill.hpp (hunk 2 above)
Comment (apmccartney) [reply src/RECONR/ReferenceGrid/src/fill.hpp]: \<grey beard professor\> obviously the thirteenth root of 10 \</grey beard professor\>

comment incoming

File: src/RECONR/ReferenceGrid/src/fill.hpp (hunk 6 above)
Comment (apmccartney) [reply src/RECONR/ReferenceGrid/src/fill.hpp]: let's unpack it one step at a time.

```cpp
[ =, state = lowerBound ]() mutable { return state *= thresholdRatio; }
```

This is a mutable lambda. Each time it's invoked, it generates a different value, i.e.
```
call -> lowerBound
call -> lowerBound * thresholdRatio
call -> lowerBound * thresholdRatio * thresholdRatio
call -> lowerBound * thresholdRatio * thresholdRatio * thresholdRatio
...
```

`ranges::view::generate` takes a callable objects and returns an infinite (lazy) range of values from repeatedly calling it's parameter object, e.g.

```
[ lowerBound,
  lowerBound * thresholdRatio,
  lowerBound * thresholdRatio * thresholdRatio,
  lowerBound * thresholdRatio * thresholdRatio * thresholdRatio,
  ... ]
```

`ranges::view::take_while` takes values from an input range until a predicate fails. In this case, it takes values while those values are less than the upper bound.

`ranges::action::push_back` is appending each value in the range to buffer.

File: src/RECONR/ReferenceGrid/test/call.test.cpp (hunk 7 above)
Comment (apmccartney) [reply src/RECONR/ReferenceGrid/test/call.test.cpp]: accidental

**Pull request merged**